### ISSUE TITLE
CoreText produces unexpected results when laying out text containing tabs

### ISSUE THREAD

**Issue by cmyr** [linebender/piet#413]
I've only tested this so far on 10.14 and 10.15, so it's possible it's improved on big sur, but I've run into a number of problems when my text layouts include tabs; they seem to cause the layout engine to produce invalid lines. This issue seems to occur whenever the layout in question contains multiple lines, either due to soft or hard breaks.

I've gone so far as to open my annual radar, number `FB9036971`.

There may be some possible hacky work-arounds that I haven't explored yet, such as removing all the default `CTTextTab` stops from the paragraph layout, and setting [`kCTParagraphStyleSpecifierDefaultTabInterval`](https://developer.apple.com/documentation/coretext/ctparagraphstylespecifier/kctparagraphstylespecifierdefaulttabinterval?language=objc) to zero, but that feels gross. Of course the current situation also feels gross.

I think for the time being I will suppress the insertion of tabs on macOS in druid, or replace them with spaces or something?

cc @jjgod, not sure if you have any insight here?

reproduction in a swift playground:

```swift
import AppKit
import PlaygroundSupport

let otherExample = "a\t\t\t\nbb"

// Changing this to CGFloat.greatestFiniteMagnitude is also broken
let widthConstraint: CGFloat = 100;

class MyView: NSView {
    let myString = NSAttributedString(string: otherExample)

    override func draw(_ rect: CGRect) {
        let ctx = NSGraphicsContext.current!.cgContext
        NSColor.white.setFill()
        ctx.fill(rect)
        let framesetter = CTFramesetterCreateWithAttributedString(myString)
        let charRange = CFRange(location: 0, length: myString.length);


        let layoutConstraints = CGSize(width: widthConstraint, height: CGFloat.greatestFiniteMagnitude)
        let frameSize = CTFramesetterSuggestFrameSizeWithConstraints(framesetter, charRange, nil, layoutConstraints, nil);

        let frameRect = CGRect(origin: CGPoint.zero, size: frameSize);
        let frame = CTFramesetterCreateFrame(framesetter, charRange, CGPath(rect: frameRect, transform: nil), nil)

        let lines = CTFrameGetLines(frame) as! [CTLine]
        debugLines(lines: lines)

        ctx.saveGState()

        ctx.translateBy(x: 0.0, y: frameRect.height)
        ctx.scaleBy(x: 1.0, y: -1.0)
        ctx.translateBy(x: 20.0, y: -20.0)

        CTFrameDraw(frame, ctx)
        ctx.restoreGState()

    }

    override var isFlipped: Bool {
        true
    }
}

func debugLines(lines: [CTLine]) {
    for (i, line) in lines.enumerated() {
        print("\(i): \(CTLineGetStringRange(line))")
    }
}

// Present the view in Playground
var view = MyView(frame: NSRect(x: 0, y: 0, width: 400, height: 400));
PlaygroundPage.current.liveView = view

```

**Comment by raphlinus** [comment]
I ran it on Bug Sur, not in the Playground (couldn't figure out how to do that), but as a Cocoa app in XCode. It displays just "a" and logs the following:

```
0: CFRange(location: 0, length: 2)
1: CFRange(location: 2, length: 3)
```

One trailing tab, displays "a" and "bb" and logs this:

```
0: CFRange(location: 0, length: 3)
1: CFRange(location: 3, length: 2)
```

Seems like an interesting bug deep in Apple's text code.

**Comment by cmyr** [comment]
I'm not ruling out that I'm just using it wrong, but it sure feels funny.

**Comment by jjgod** [comment]
While there might be something tricky in tabstop handling (which we can look into further), you are not supposed to pass the output of `CTFramesetterSuggestFrameSizeWithConstraints` back to `CTFramesetterCreateFrame()`, there is no guarantee that will give you the same result.

Just pass the original rect size as the input to create the `CGPath` for `CTFramesetterCreateFrame()` and you will be fine.

**Comment by cmyr** [comment]
@jjgod 😳I thought this was explicitly what `CTFramesetterSuggestFrameSizeWithConstraints` was for. In any case, this is good to know, thank you!

I don't have an input rect; I have input constraints, and I'm trying to lay out text that fits those constraints. I think maybe the conclusion here is that I should be using `CTTypesetter` directly.

**Comment by cmyr** [comment]
I'll play around a bit. If I pass infinite or float.max constraints I'll end up still having to go and get the lines, compute their actual bounds manually, and position them manually (I can't use `CTFrameGetLineOrigins` because, for instance, with a very large width RTL text will be positioned at the max edge of my constraints) and I also can't trust the rect returned by `CTFramesetterSuggestFrameSizeWithConstraints` then I think I'm no longer getting any benefit from trying to use `CTFramesetter`.

In any case this has been very helpful, I have clear directions to continue experimenting in, and it's much appreciated.

If you think it would be helpful for me to open radars against missing docs I'd be happy to do that.

**Comment by jjgod** [comment]
Sure, we can always use better documentation.

I should make clear that `CTFramesetterSuggestFrameSizeWithConstraints` is doing the layout once and `CTFramesetterCreateFrame` once as well. So even just from a performance point of view you should avoid the double layout pass. So it sounds like getting the lines and line origins from the resulting frame would probably worth the trouble.

**Comment by cmyr** [comment]
This was resolved in #407
